Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+    F('g', 2)                  \
+    F('G', 2)                  \
+    F('h', 2)                  \
+    F('H', 2)                  \
+    F('i', 1)                  \
+    F('l', 2)                  \
+    F('n', 2)                  \
+    F('N', 2)                  \
+    F('p', 2)                  \
+    F('P', 2)                  \
+    F('q', 1)                  \
+    F('r', 1)                  \
+    F('s', 2)                  \
+    F('t', 2)                  \
+    F('w', 2)                  \
+    F('x', 2)                  \
+    F('y', 2)                  \
+    F(':', 0)                  \
+    F('=', 1)                  \
+    F('#', 0)
+
+enum class AddressType {
+    Unset,
+    Line,
+    LastLine,
+    ContextAddress,
+};
+
+class Address {
+public:
+    Address() = default;
+
+    explicit Address(size_t line)
+        : m_line_number(line)
+        , m_address_type(AddressType::Line)
+    {
+    }
+
+    explicit Address(AddressType address_type)
+        : m_address_type(address_type)
+    {
+        VERIFY(address_type == AddressType::LastLine || address_type == AddressType::ContextAddress);
+    }
+
+    size_t line_number() const
+    {
+        VERIFY(m_address_type == AddressType::Line);
+        return m_line_number;
+    }
+
+    AddressType address_type() const { return m_address_type; }
+
+    bool matches(StringView /*pattern_space*/, size_t line_number, bool is_last_line) const
+    {
+        switch (m_address_type) {
+        case AddressType::Line:
+            return line_number == m_line_number;
+        case AddressType::LastLine:
+            return is_last_line;
+        default:
+            Core::File::standard_error()->write(String::formatted("Addressing type not implemented: {}", int(m_address_type)));
+            return false;
+        }
+    }
+
+private:
+    size_t m_line_number { 0 };
+    AddressType m_address_type { AddressType::Unset };
+};
+
+namespace AK {
+template<>
+class Formatter<Address> : public StandardFormatter {
+public:
+    void format(FormatBuilder& format_builder, Address address)
+    {
+        auto& builder = format_builder.builder();
+        switch (address.address_type()) {
+        case AddressType::Line:
+            builder.appendff("{}", address.line_number());
+            break;
+        case AddressType::LastLine:
+            builder.append('$');
+            break;
+        case AddressType::ContextAddress:
+            VERIFY_NOT_REACHED();
+        case AddressType::Unset:
+            break;
+        }
+    }
+};
+}
+
+struct Command {
+    Address addr1;
+    Address addr2;
+    char function = '\0';
+    StringView arguments;
+
+    void enable_for(StringView pattern_space, size_t line_number, bool is_last_line)
+    {
+        m_is_enabled = selects(pattern_space, line_number, is_last_line);
+    }
+
+    bool is_enabled() const { return m_is_enabled; }
+
+private:
+    bool m_is_enabled { false };
+    bool m_is_selection_active { false };
```

Comment:
Please put these at the end of the struct definition (after the `selects()` member function).

Reply:
Done